[NCGenerics] Even more fixes for building the stdlib (Part 3)

kavon opened the pull request:
fixes rdar://120462547&120512544

Just another assorted set of changes as I fix things while getting the stdlib building with NoncopyableGenerics.

kavon commented:
@swift-ci smoke test

kavon commented:
@swift-ci please smoke test Linux

kavon committed:
[NCGenerics] fix Copyable/Escapable declarations

When building the stdlib with noncopyable generics enabled,
requirements for Copyable and Escapable  end up being inferred on the
protocols Copyable and Escapable, themselves. So we need to opt-out
Copyable from requiring Escapable, and vice versa.

kavon commented:
@swift-ci please smoke test Linux

kavon committed:
[NCGenerics] add coverage for `any Error`

kavon committed:
add Linux/macOS builder NoncopyableGenerics presets

kavon commented:
preset=buildbot_linux_ncg
@swift-ci please test linux

kavon committed:
[NCGenerics] infinite substitution hack

Workaround for rdar://119950540 when dealing with invertible protocols.

This should be sound because "its an invariant that subst() on a
conformance returns the same thing as a global lookup with subst() of
the conforming type, assuming you don't have overlapping conformances."

kavon committed:
[SILGen] conformsToInvertible of no-implicit-copy

The "no implicit copy" parameters such as a `consuming` or `borrowing`
parameter of an otherwise Copyable type was not being handled correctly
by `conformsToInvertible`. We don't have conformances for SIL types in
the AST, so they're supposed to be handled earlier.

fixes rdar://120462547

kavon committed:
[NCGenerics] disable mangling roundtrip checking

This is a temporary measure until I get a chance to fix the symbol
mangling. This round-tripper is exactly flagging the current
known issue in that all of the symbols get extra `Copyable` and
`Escapable` requirements mangled into them, and thus reconsituting the
type from the mangled name doesn't give you the same type back.

kavon committed:
[NFC] fast-path to skip a requirement scan

kavon committed:
[NCGenerics] printing class-constrained generics

In cases where the generic parameter is class-constrained,
`GenericSignature::requiresProtocol` will not contain `Copyable` or
`Escapable` because GenericSignature minimization will recognize that
the class already requires them.

Thus, because classes always require those protocols, we can
simply ask if the generic parameter is required to be a class to
determine if it had any inverses.

kavon commented:
@swift-ci please test

slavapestov commented on lib/AST/ProtocolConformanceRef.cpp:
```diff
@@ -118,6 +118,21 @@ ProtocolConformanceRef::subst(Type origType, InFlightSubstitution &IFS) const {
+      auto conf =
+          proto->getModuleContext()->lookupConformance(substType, proto);
+
+      if (conf)
```
You can just `return conf`

slavapestov commented on lib/IRGen/GenExistential.cpp:
```diff
@@ -1858,8 +1858,11 @@ OwnedAddress irgen::emitBoxedExistentialContainerAllocation(IRGenFunction &IGF,
+  // Should only be one stored conformance, for the Error protocol.
+  // But there are 1+NumInvertibleProtocols conformances in the existential's
+  // layout.
+  assert(conformances.size() == 1+NumInvertibleProtocols);
```
I feel like this is a bug because Error's requirement signature should have `Self: Copyable`. So the minimal signature for `<T where T: Copyable, T: Error>` is just `<T where T: Error>`, and ExistentialLayout::getProtocols() should then just return `Error`.

kavon replied:
I've fixed this so that it just has one protocol, Error, in the layout, since Copyable is implied.

slavapestov commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -1472,6 +1472,10 @@ static void reconstituteInverses(GenericSignature genericSig,
     assert(tp);
 
+    // Any generic parameter requiring a class could not have an inverse.
+    if (genericSig->requiresClass(tp))
```
does `<T where T: AnyObject, T: Copyable>` minimize as `<T where T: AnyObject>`? It should via a concrete conformance but I believe that needs to be implemented in the requirement machine itself.

slavapestov replied:
Maybe instead of doing `sig->requiresProtocol(paramTy, Copyable)` you can map paramTy into the signature's environment and do a global lookup with the archetype to Copyable? Then you won't need this special case.

kavon committed:
[NCGenerics] ensure ExistentialLayout is minimal

Instead of injecting Copyable & Escapable protocols into every
ExistentialLayout, only add those protocols if the existing protocols
don't already imply them. This simplifies things like `any Error`
protocol, so it still only lists one protocol in its existential layout.
But existentials like `any NoCopyP` still end up with a Copyable in its
layout.

kavon committed:
[NCGenerics] fix AnyObject and inverses

resolves rdar://120512544

kavon commented:
@swift-ci smoke test

kavon committed:
[NFC] non-null assertion for `inheritsFrom`

kavon committed:
[NFC] ensure ErasureExpr gets valid conformances

kavon commented:
@swift-ci please test

kavon committed:
[NFC] rename `preprocessType`

kavon committed:
[NCGenerics] handle `ReferenceStorageType` queries

kavon committed:
[NCGenerics] workaround availability issue

In the Concurrency module, an unavailable Sendable extension triggers an
 error about unavailable Copyable conformance. This conformance came
 through in the SubstitutionMap of a BoundGenericType, so I think it's
 ok to just ignore these.

kavon committed:
[NFC] forEachMissingConformance doesn't use module

kavon committed:
[NFC] combine `collectExistentialConformances`

There are a bunch of static `collectExistentialConformances` copied
around Sema and SILGen that are almost the same, save for whether they
want to permit missing conformances and/or check conditional
conformances.

This commit requestifies and combines all but one of these functions
into a `ModuleDecl::collectExistentialConformances`. The motivation for
this clean-up is another place that will need this procedure.

kavon commented:
@swift-ci please smoke test

kavon committed:
[NCGenerics] fix `Builtin.createAsyncTask`

The SILGen emission for `Builtin.createAsyncTask` performs an
existential erasure and assumed that `any Any.Type` has no protocols in
its ExistentialLayout, but it's now possible to have some there with
NoncopyableGenerics.

kavon commented:
@swift-ci please test
